service/history/workflow/task_generator.go

```diff
@@ -185,21 +189,30 @@ func (r *TaskGeneratorImpl) GenerateWorkflowCloseTasks(
 		// or execution info in mutable state. For immediate task, visibility timestamp
 		// should always be when the task is generated so that task_latency_queue/load
 		// truly measures only task processing/loading latency.
-		closeTime := timestamp.TimeValue(closeEvent.GetEventTime())
 		closeTasks = append(closeTasks,
 			&tasks.CloseExecutionVisibilityTask{
 				// TaskID, VisibilityTimestamp is set by shard
 				WorkflowKey: r.mutableState.GetWorkflowKey(),
 				Version:     currentVersion,
 			},
-			&tasks.DeleteHistoryEventTask{
+		)
+		closeTime := timestamp.TimeValue(closeEvent.GetEventTime())
+		if r.config.DurableArchivalEnabled() {
+			closeTasks = append(closeTasks, &tasks.ArchiveExecutionTask{
+				// TaskID is set by shard
+				WorkflowKey:         r.mutableState.GetWorkflowKey(),
+				VisibilityTimestamp: closeTime,
```

Comment:
@yycptt to confirm. I think this will get override in shard layer? To persist the close time, it might need a new field?

Reply:
Yes, for scheduled tasks, visibility timestamp is the actual task fire time, so it should be the desired task execution time. For immediate tasks, visibility timestamp is purely for metrics purpose and should always be task generation time. To make things easier, shard info will overwrite VisibilityTimestamp for all immediate tasks to be time.now(). 

DeleteHistoryEventTask is a scheduled task, so it's using the close time. Archival task is an immediate task I remember, so you can leave this field empty. 

I don't think archival task needs the close time. Even if it does, close time can be retrieved from mutable state when task is being processed.

Reply:
Ok cool. I removed the visibility timestamp from the proto, and I added an execution close time there instead.

Reply:
No, don't remove the visibility timestamp from proto. The timestamp still needs to be persisted to calculate the delay when processing the task. 

Also, I am confused why execution close time needs to be persisted in the task. If it's required for executing the archival task, its value can be retrieved from the mutable state I think.

Reply:
It is required, but I think storing it makes things easier because we don't need to access mutable state this way. I'll persist the mutable state though.